tridiagonal/include/tridiagonal.inc

```diff
@@ -76,104 +27,80 @@ contains
-  if(allocated(cc)) deallocate(cc)
-  !$OMP END SINGLE !< There is an implicit barrier.
-
-return
-end subroutine close_tridiagonal
-
-!----------------------------------------------------------------
-
-end module tridiagonal_mod
-
-! <INFO>
-
-!   <BUG>
-!     Optional arguments A,B,C have no intent declaration,
-!     so the default intent is inout. The value of A(N) is modified
-!     on output, and B and C are unchanged.
-!   </BUG>
-!   <NOTE>
-!       The following private allocatable arrays save the relevant information
-!  if one recalls tri_invert without changing (A,B,C):
-!  <PRE>
-!        allocate ( e  (size(x,1), size(x,2), size(x,3)) )
-!        allocate ( g  (size(x,1), size(x,2), size(x,3)) )
-!        allocate ( cc (size(x,1), size(x,2), size(x,3)) )
-!        allocate ( bb (size(x,1), size(x,2)) )
-! </PRE>
-!  This storage is deallocated when close_tridiagonal is called.
-!   </NOTE>
-!   <FUTURE>
-!     Maybe a cleaner version?
-!   </FUTURE>
-
-! </INFO>
-!> @}
-! close documentation grouping
+!!
+!! For mixed precision, this routine uses the kind size macro(FMS_TRID_KIND_) to determine
+!! which module variables are used/stored. This means a,b, and c values will only be stored for calls
+!! of the same real kind value unless store_both_kinds is present and .true..
+subroutine TRI_INVERT_(x,d,a,b,c, store_both_kinds)
+
+    real(FMS_TRID_KIND_), intent(out), dimension(:,:,:) :: x !< Solution to the tridiagonal system of equations
+    real(FMS_TRID_KIND_), intent(in),  dimension(:,:,:) :: d !< The right-hand side term, see the schematic above.
+    real(FMS_TRID_KIND_), optional,    dimension(:,:,:) :: a,b,c !< Left hand side terms(see schematic on module page).
+                                                !! If not provided, values from last call are used
+    logical, optional                                   :: store_both_kinds !< Will save module state
+                                                         !! variables for both kind types in order to be used in
+                                                         !! subsequent calls with either kind.
+
+    real(FMS_TRID_KIND_), dimension(size(x,1),size(x,2),size(x,3)) :: f
+    integer, parameter :: kindl = FMS_TRID_KIND_
+
+    integer :: k
+
+    if(present(a)) then
+        !$OMP SINGLE
+        INIT_VAR = .true.
+        if(allocated(TRID_REAL_TYPE%e))     deallocate(TRID_REAL_TYPE%e)
+        if(allocated(TRID_REAL_TYPE%g))     deallocate(TRID_REAL_TYPE%g)
+        if(allocated(TRID_REAL_TYPE%bb))    deallocate(TRID_REAL_TYPE%bb)
+        if(allocated(TRID_REAL_TYPE%cc))    deallocate(TRID_REAL_TYPE%cc)
+        allocate(TRID_REAL_TYPE%e (size(x,1),size(x,2),size(x,3)))
+        allocate(TRID_REAL_TYPE%g (size(x,1),size(x,2),size(x,3)))
+        allocate(TRID_REAL_TYPE%bb(size(x,1),size(x,2)))
+        allocate(TRID_REAL_TYPE%cc(size(x,1),size(x,2),size(x,3)))
+        !$OMP END SINGLE
+
+        TRID_REAL_TYPE%e(:,:,1) = - a(:,:,1) / b(:,:,1)
```

Comment:
its just the way this module is written. It's kind of like instead of an init routine you provide the first call to `tri_invert` with a,b, and c, and then those values are stored and used when the optional arguments aren't provided. It assumes a,b, and c will all be there or not at all.

I'll add `present` checks for b and c as well just to be safe.